In [1]:
import pandas as pd

# Chargement de la base de données contenant les infos + performances des joueurs

df = pd.read_csv('./data/merged_data.csv')

print(f"Dataset chargé: {len(df)} lignes, {len(df.columns)} colonnes")

Dataset chargé: 10285 lignes, 78 colonnes


# 2 - Nettoyage et pre-processing de la BDD

Grâce au scraping mis en place, on récupère un jeu de données assez conséquant (10k+ données et presque 80 colonnes). Afin de réduire le volume de données à traîter pour accélérer le temps de calcul des modèles sans perdre trop d'information, on peut commencer par supprimer les données très corrélées (>= à 80% en valeur absolue). Commençons d'abord par calculer le nombre de paires de variables qui sont largement corrélées :

In [2]:
from clean_dataset_fun import analyze_correlations

# Analyse des corrélations élevées (>= 80%)

correlations_list = analyze_correlations(df, threshold=0.80, verbose=True)

Variables avec une corrélation >= 80% (en valeur absolue):

Nombre total de paires trouvées: 172

  1. NBA_FANTASY_PTS_RANK <-> WNBA_FANTASY_PTS_RANK : +0.9957
  2. WNBA_FANTASY_PTS     <-> NBA_FANTASY_PTS      : +0.9949
  3. FGM_RANK             <-> PTS_RANK             : +0.9924
  4. FGM                  <-> PTS                  : +0.9895
  5. FG3A                 <-> FG3M                 : +0.9857
  6. MIN_RANK             <-> MIN                  : -0.9835
  7. FGA_RANK             <-> PTS_RANK             : +0.9834
  8. FGA                  <-> PTS                  : +0.9821
  9. FTA_RANK             <-> FTM_RANK             : +0.9821
 10. FTM                  <-> FTA                  : +0.9821
 11. W_RANK               <-> W                    : -0.9816
 12. FGM                  <-> FGA                  : +0.9793
 13. FGM_RANK             <-> FGA_RANK             : +0.9780
 14. adjusted_salary      <-> Salary               : +0.9779
 15. FG3A_RANK            <-> FG3M_RANK        

On remarque qu'il y a 172 paires de variables qui sont extrêmement corrélées. On va donc supprimer une partie de ces variables qui portent vraissemblablement la même information que les autres (tout en "protégeant certaines colonnes pour qu'elles ne soient pas supprimées, comme celles du salaire car elles seront les targets de nos modèles).

In [3]:
from clean_dataset_fun import remove_highly_correlated_features

# Suppression des colonnes hautement corrélées, en protégeant certaines colonnes qu'on souhaite conserver

protected_columns = ['Salary', 'adjusted_salary', 'next_adjusted_salary']

columns_to_drop = remove_highly_correlated_features(df, threshold=0.80, protected_cols=protected_columns, verbose=False)

df_cleaned = df.drop(columns=columns_to_drop)

print(f"Dataset chargé: {len(df)} lignes, {len(df.columns)} colonnes")

print(f"Dataset nettoyé: {len(df_cleaned)} lignes, {len(df_cleaned.columns)} colonnes")

Dataset chargé: 10285 lignes, 78 colonnes
Dataset nettoyé: 10285 lignes, 42 colonnes


En procédant de cette manière, on arrive à diviser le nombre de variables par 2, ce qui va considéramment améliorer la performance. Maintenant, concentrons nous sur la suppressions des outliers de cette nouvelle base de données. En effet, il y a certaines données qui vont compromettre la qualité de nos modèles car elles sont associées à des cas très particuliers. Par exemple, on va supprimer les informations sur les joueurs dont le salaire ajusté à l'inflation est inférieur 500k$, car c'était le salaire minimum en NBA en 2000. Les joueurs gagnant un montant inférieur ne jouent pas en NBA de manière régulière ou à plein temps, on supprime donc les points associés. De plus, il arrive qu'un joueur voit son salaire drastiquement diminuer car il arrive en fin de carrière, or nous on s'intéresse plutôt à l'augmentation ou baisse des salaires des joueur actifs, en fonction de leur qualité de jeu. On supprime donc les points associés à une baisse considérable de salaire qui indique généralement une fin de carrière (ainsi que les points suivants car un joueur peut jouer quelques saisons avec un salaire de fin de carrière).

In [4]:
from clean_dataset_fun import clean_outliers

df_clean = clean_outliers(df_cleaned, min_salary=500000, drop_threshold=-0.70)

NETTOYAGE DES OUTLIERS

Dataset initial: 10285 lignes

1. Suppression des salaires ajustés < 500,000$
   Lignes supprimées: 404
   Lignes restantes: 9881

2. Détection des fins de carrière (baisse > 70%)
   Joueurs avec baisse > 70%: 343
   Lignes identifiées (fin de carrière): 804
   Lignes supprimées (fin de carrière): 804
   Lignes restantes: 9077

RÉSUMÉ DU NETTOYAGE
Dataset initial:                10285 lignes
Supprimés (salaire < 500,000$):    404 lignes
Supprimés (fin de carrière):    804 lignes
Total supprimé:                 1208 lignes (11.75%)
Dataset final:                  9077 lignes

 Nettoyage des outliers terminé


De cette manière, on supprime plus de 1000 données qui vont contribuer à dégrader la qualité de nos modèles. Maintenant que nous avons supprimé les variables et points de mauvaise qualité, on va appliqué un pre-processing du dataset en plusieurs étapes : 

1. Suppression des colonnes inutiles pour les modèles ('PLAYER_ID', 'PLAYER_NAME'...)
2. Encodage des variables catégorielles (comme la position des joueurs, ID de l'équipe)
3. Normalisation des variables numériques
4. Vérification des valeurs manquantes (on supprime les données qui n'ont pas de 'next_adjusted_salary', car c'est la variable target de nos modèles. Sinon on remplace par la valeur moyenne)

In [5]:
from clean_dataset_fun import preprocess_pipeline

df_preprocessed, artifacts = preprocess_pipeline(df_clean, verbose=True)

print(f"Dataset preprocessed: {len(df_preprocessed)} lignes, {len(df_preprocessed.columns)} colonnes")

PIPELINE DE PRÉTRAITEMENT DES DONNÉES

1. Dataset initial: (9077, 42)

2. Colonnes supprimées: ['PLAYER_ID', 'PLAYER_NAME', 'NICKNAME', 'TEAM_ID', 'Team', 'Salary', 'Season', 'Year', 'adjusted_salary', 'Rank']
Nouvelles dimensions: (9077, 32)

3. Encodage des variables catégorielles:

Position encodée:
   Nombre de classes: 9
   Classes: ['C', 'F', 'G', 'GF', 'PF', 'PG', 'SF', 'SG', 'nan']
   Mapping: {'C': np.int64(0), 'F': np.int64(1), 'G': np.int64(2), 'GF': np.int64(3), 'PF': np.int64(4), 'PG': np.int64(5), 'SF': np.int64(6), 'SG': np.int64(7), 'nan': np.int64(8)}

TEAM_ABBREVIATION encodée:
   Nombre de classes: 36

Colonnes catégorielles originales supprimées: ['Position', 'TEAM_ABBREVIATION']

4. Colonnes à normaliser (31):
   ['AGE', 'W', 'L', 'FG_PCT', 'FG3M', 'FG3_PCT', 'FTM', 'OREB', 'DREB', 'AST']...

Normalisation effectuée avec StandardScaler
   Moyenne ≈ 0, Écart-type ≈ 1

5. Valeurs manquantes détectées:
   next_adjusted_salary: 1428 (15.73%)
   Changed_team: 1428 (15.7

# 3 - Modélisation

## 3.2 RadomForestRegressor

In [6]:
from train_random_forest_model import train_random_forest, get_feature_importance

model_dict = train_random_forest(df_preprocessed, target_column='next_adjusted_salary')

get_feature_importance(model_dict, top_n=5)

Taille du jeu d'entraînement: 5354
Taille du jeu de test: 2295

Entraînement du modèle RandomForestRegressor...

RÉSULTATS DU MODÈLE

Performance sur l'ensemble d'ENTRAÎNEMENT:
   R² Score:  0.9562
   RMSE:      $2,006,489.84
   MAE:       $1,368,615.04

Performance sur l'ensemble de TEST:
   R² Score:  0.7034
   RMSE:      $5,261,600.50
   MAE:       $3,596,636.42

Top 5 features les plus importantes:
feature  importance
    FTM    0.297834
    YOE    0.166158
   DREB    0.077959
    PFD    0.040127
   FG3M    0.037834


feature  importance
0                         FTM    0.297834
1                         YOE    0.166158
2                        DREB    0.077959
3                         PFD    0.040127
4                        FG3M    0.037834
5                    TOV_RANK    0.032708
6                         AGE    0.029541
7                         AST    0.022266
8            Position_encoded    0.021681
9                    TD3_RANK    0.021528
10            PLUS_MINUS_RANK    0.020253
11                     FG_PCT    0.018968
12                   PFD_RANK    0.017715
13                          L    0.016710
14                FT_PCT_RANK    0.016332
15                    GP_RANK    0.015227
16                   DD2_RANK    0.015070
17                         PF    0.012845
18                 TEAM_W_PCT    0.012354
19               FG3_PCT_RANK    0.012344
20                        BLK    0.012156
21  TEAM_ABBREVIATION_encoded    0.012092
22                          W    0.011058
23                        STL    0.010953
24                       OREB    0.010377
25                       BLKA    0.009553
26                    FG3_PCT    0.008690
27                        DD2    0.008291
28               Changed_team    0.006629
29                        TD3    0.003263
30                 TEAM_COUNT    0.001482